libimagtodo/src/task.rs

```diff
@@ -1,34 +1,144 @@
+use std::result::Result as RResult;
+
 use toml::Value;
+use uuid::Uuid;
 
 use task_hookrs::task::Task as TTask;
+use task_hookrs::import::{import_task, import_tasks};
 
 use libimagstore::store::{FileLockEntry, Store};
-use libimagstore::storeid::IntoStoreId;
+use libimagstore::storeid::{IntoStoreId, StoreIdIterator, StoreId};
 use module_path::ModuleEntryPath;
 
-use error::{TodoError, TodoErrorKind};
+use error::{TodoError, TodoErrorKind, MapErrInto};
 use result::Result;
 
 /// Task struct containing a `FileLockEntry`
 #[derive(Debug)]
-pub struct Task<'a> {
-    pub flentry : FileLockEntry<'a>,
-}
+pub struct Task<'a>(FileLockEntry<'a>);
 
 impl<'a> Task<'a> {
+
     /// Concstructs a new `Task` with a `FileLockEntry`
-    pub fn new(fle : FileLockEntry<'a>) -> Task<'a> {
-        Task {
-            flentry : fle
-        }
+    pub fn new(fle: FileLockEntry<'a>) -> Task<'a> {
+        Task(fle)
+    }
+
+    pub fn import<R: BufRead>(store: &'a Store, mut r: R) -> Result<(Task<'a>, String, Uuid)> {
+        let mut line = String::new();
+        r.read_line(&mut line);
+        import_task(&line.as_str())
+            .map_err_into(TodoErrorKind::ImportError)
+            .and_then(|t| {
+                let uuid = t.uuid().clone();
+                t.into_task(store).map(|t| (t, line, uuid))
+            })
+    }
+
+    /// Get a task from an import string. That is: read the imported string, get the UUID from it
+    /// and try to load this UUID from store.
+    ///
+    /// Possible return values are:
+    ///
+    /// * Ok(Ok(Task))
+    /// * Ok(Err(String)) - where the String is the String read from the `r` parameter
+    /// * Err(_)          - where the error is an error that happened during evaluation
+    ///
+    pub fn get_from_import<R: BufRead>(store: &'a Store, mut r: R) -> Result<RResult<Task<'a>, String>>
+    {
+        let mut line = String::new();
+        r.read_line(&mut line);
+        Task::get_from_string(store, line)
+    }
+
+    /// Get a task from a String. The String is expected to contain the JSON-representation of the
+    /// Task to get from the store (only the UUID really matters in this case)
+    ///
+    /// For an explanation on the return values see `Task::get_from_import()`.
+    pub fn get_from_string(store: &'a Store, s: String) -> Result<RResult<Task<'a>, String>> {
+        import_task(s.as_str())
+            .map_err_into(TodoErrorKind::ImportError)
+            .map(|t| t.uuid().clone())
+            .and_then(|uuid| Task::get_from_uuid(store, uuid))
+            .and_then(|o| match o {
+                None    => Ok(Err(s)),
+                Some(t) => Ok(Ok(t)),
+            })
+    }
+
+    /// Get a task from an UUID.
+    ///
+    /// If there is no task with this UUID, this returns `Ok(None)`.
+    pub fn get_from_uuid(store: &'a Store, uuid: Uuid) -> Result<Option<Task<'a>>> {
+        let store_id = ModuleEntryPath::new(format!("taskwarrior/{}", uuid)).into_storeid();
```

Comment:
> "taskwarrior/{}" is used quite often, and if it is changed, all occurences would need to be changed. 

Indeed, but I want that extra level for beeing able to add other backends later on. For example `"tudu/{}"` (tudu is another todo tool) would be an option.

As we do not (yet) have a helper for creating multiple backends, this way is how it has to be, IMHO.

Doing abstraction over ModuleEntryPath and so on, as you suggested, would be part of a backends-helper crate... by now this will suffice IMO.